component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/handlers/AbstractMTLSTokenBindingGrantHandler.java

```diff
@@ -0,0 +1,111 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls.handlers;
+
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jose.util.X509CertUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.model.HttpRequestHeader;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls.utils.CommonConstants;
+
+import java.io.ByteArrayInputStream;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+public class AbstractMTLSTokenBindingGrantHandler {
+
+    private static final Log log = LogFactory.getLog(MTLSTokenBindingAuthorizationCodeGrantHandler.class);
+
+    public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx, boolean validateScope) throws IdentityOAuth2Exception {
+
+        // Get MTLS certificate from transport headers.
+        HttpRequestHeader[] requestHeaders = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getHttpRequestHeaders();
+        String headerName = IdentityUtil.getProperty(CommonConstants.MTLS_AUTH_HEADER);
+
+        Optional<HttpRequestHeader> certHeader =
+                Arrays.stream(requestHeaders).filter(httpRequestHeader ->
+                        headerName.equals(httpRequestHeader.getName())).findFirst();
+
+        String authenticatorType = (String) tokReqMsgCtx.getOauth2AccessTokenReqDTO().getoAuthClientAuthnContext()
+                .getParameter(CommonConstants.AUTHENTICATOR_TYPE_PARAM);
+        if (certHeader.isPresent() && CommonConstants.AUTHENTICATOR_TYPE_MTLS.equals(authenticatorType)) {
+            Base64URL certThumbprint = null;
+            if (log.isDebugEnabled()) {
+                log.debug("Client MTLS certificate found: " + certHeader);
+            }
+            try {
+                if (certHeader.get().getValue() != null) {
+                    X509Certificate certificate = parseCertificate(certHeader.get().getValue()[0]);
+                    certThumbprint = X509CertUtils.computeSHA256Thumbprint(certificate);
+                }
+            } catch (CertificateException e) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Error occurred while calculating the thumbprint of the client MTLS certificate", e);
```

Comment:
Add fullstop at the end of message. Check other places as well.

Reply:
Fixed 479e087